File: src/repository/Requests.go
package repository

import (
	"log"

	"github.com/averageflow/goscope/src/types"
	"github.com/averageflow/goscope/src/utils"
)

// FetchDetailedRequest fetches all details from a request via its UUID.
func FetchDetailedRequest(requestUID string) types.DetailedRequest {
	var body string

	var headers string

	var result types.DetailedRequest

	row := QueryDetailedRequest(utils.DB, requestUID)

	err := row.Scan(
		&result.UID,
		&result.ClientIP,
		&result.Method,
		&result.Path,
		&result.URL,
		&result.Host,
		&result.Time,
		&headers,
		&body,
		&result.Referrer,
		&result.UserAgent,
	)
	if err != nil {
		log.Println(err.Error())
	}

	result.Body = utils.PrettifyJSON(body)
	result.Headers = utils.PrettifyJSON(headers)

	return result
}

// FetchDetailedResponse fetches all details of a response via its UUID.
func FetchDetailedResponse(responseUUID string) types.DetailedResponse {
	var body string

	var headers string

	var result types.DetailedResponse

	row := QueryDetailedResponse(utils.DB, responseUUID)

	err := row.Scan(
		&result.UID,
		&result.ClientIP,
		&result.Status,
		&result.Time,
		&body,
		&result.Path,
		&headers,
		&result.Size,
	)
	if err != nil {
		log.Println(err.Error())
	}

	result.Body = utils.PrettifyJSON(body)
	result.Headers = utils.PrettifyJSON(headers)

	return result
}

// FetchRequestList fetches a list of summarized requests.
func FetchRequestList(offset int) []types.SummarizedRequest {
	var result []types.SummarizedRequest

	rows, err := QueryGetRequests(utils.DB, offset)
	if err != nil {
		log.Println(err.Error())

		return result
	}

	if rows.Err() != nil {
		log.Println(rows.Err().Error())

		return result
	}

	defer rows.Close()

	for rows.Next() {
		var request types.SummarizedRequest

		err := rows.Scan(
			&request.UID,
			&request.Method,
			&request.Path,
			&request.Time,
			&request.ResponseStatus,
		)
		if err != nil {
			log.Println(err.Error())
			return result
		}

		result = append(result, request)
	}

	return result
}

// FetchSearchRequests fetches a list of summarized requests that match the input parameters of search.
func FetchSearchRequests(search string, filter *types.RequestFilter, offset int) []types.SummarizedRequest {
	var result []types.SummarizedRequest

	rows, err := QuerySearchRequests(utils.DB, utils.Config.GoScopeDatabaseType, search, filter, offset)
	if err != nil {
		log.Println(err.Error())
		return result
	}

	if rows.Err() != nil {
		log.Println(rows.Err().Error())

		return result
	}

	defer rows.Close()

	for rows.Next() {
		var request types.SummarizedRequest

		err := rows.Scan(
			&request.UID,
			&request.Method,
			&request.Path,
			&request.Time,
			&request.ResponseStatus,
		)

		if err != nil {
			log.Println(err.Error())
		}

		result = append(result, request)
	}

	return result
}

File: frontend/src/interfaces/requests.ts
export interface Requests {
  method: string;
  path: string;
  time: number;
  uid: string;
  responseStatus: number;
}

export interface RequestsEndpointResponse {
  applicationName: string;
  data: Requests[];
}

export interface DetailedResponse {
  body: string;
  clientIP: string;
  headers: string;
  path: string;
  size: number;
  status: string;
  time: number;
  requestUID: string;
  uid: string;
}

export interface DetailedRequest {
  body: string;
  clientIP: string;
  headers: string;
  host: string;
  method: string;
  path: string;
  referrer: string;
  time: number;
  uid: string;
  url: string;
  userAgent: string;
}

export interface DetailedRequestResponse {
  applicationName: string;
  data: {
    response: DetailedResponse;
    request: DetailedRequest;
  };
}

export interface FilterRequest {
  method: Method[];
  status: Status[];
}

export enum Method {
  GET = "GET",
  HEAD = "HEAD",
  POST = "POST",
  PUT = "PUT",
  DELETE = "DELETE",
  CONNECT = "CONNECT",
  OPTIONS = "OPTIONS",
  TRACE = "TRACE",
  PATCH = "PATCH"
}

export enum Status {
  INFORMATIONAL = 1,
  SUCCESS = 2,
  REDIRECTION = 3,
  CLIENT_ERROR = 4,
  SERVER_ERROR = 5
}
